services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerStateModel.java

```diff
@@ -126,35 +162,54 @@ private interface StateTransition {
+    executeStateTransitionSync(message, () -> {
       if (clusterConfig == null) {
         throw new VeniceException("No configuration exists for " + clusterName);
       }
-      String controllerName = message.getTgtName();
       LOGGER.info("{} becoming leader from standby for {}", controllerName, clusterName);
-
-      if (helixManagerInitialized()) {
-        // TODO: It seems like this should throw an exception. Otherwise the case would be you'd have an instance be
-        // leader
-        // in Helix that hadn't subscribed to any resource. This could happen if a state transition thread timed out and
-        // ERROR'd
-        // and the partition was 'reset' instead of bouncing the process.
-        LOGGER.error(
-            "Helix manager already exists for instance {} on cluster {} and received controller name {}",
-            helixManager.getInstanceName(),
-            clusterName,
-            controllerName);
-
-      } else {
-        initHelixManager(controllerName);
-        initClusterResources();
-        LOGGER.info(
-            "Controller {} with instance {} is the leader of cluster {}",
-            controllerName,
-            helixManager.getInstanceName(),
-            clusterName);
-      }
     });
+
+    /**
+     * STANDBY->LEADER state transition is fine to be synchronous and slow transition, because controller client has
+     * retry logic. However, the state transition has to be executed in order, such that if there is any unfinished
+     * state transition actions from previous round (e.g. LEADER -> FOLLOWER) for the same controller, it will be
+     * blocked until the previous transition is finished.
+     */
+    try {
+      executeStateTransitionAsync(message, () -> {
+        if (helixManagerInitialized()) {
+          // TODO: It seems like this should throw an exception. Otherwise the case would be you'd have an instance be
+          // leader
+          // in Helix that hadn't subscribed to any resource. This could happen if a state transition thread timed out
+          // and
+          // ERROR'd
+          // and the partition was 'reset' instead of bouncing the process.
+          LOGGER.error(
+              "Helix manager already exists for instance {} on cluster {} and received controller name {}",
+              helixManager.getInstanceName(),
+              clusterName,
+              controllerName);
+        } else {
+          try {
+            initHelixManager(controllerName);
+          } catch (Exception e) {
+            throw new VeniceException("Failed to initialize Helix Manager for " + controllerName, e);
+          }
+          initClusterResources();
+          LOGGER.info(
+              "Controller {} with instance {} is the leader of cluster {}",
+              controllerName,
+              helixManager.getInstanceName(),
+              clusterName);
+        }
+      }).get();
```

Comment:
We probably should have a timeout since the beginning but now is also a good time to add one. Especially now all the ST work for controller is going through this single thread: `Executors.newSingleThreadExecutor(
        new DaemonThreadFactory(String.format("Controller-ST-Worker-%s", clusterName), admin.getLogContext()));` This way if something is stuck we can still fail reasonably fast and let other unstuck controller take leadership.

Reply:
Thanks for the review @xunyin8, We had a offline discussion . The idea is that we want the controller to fail if the standby-to-leader ST takes really long time, so that other  controller can have a chance to become the leader. I think it's a reasonable idea and I will update it. One clarification is that we have one ExecutorService per-cluster, so a slow transition is cluster 1 wouldn't impact cluster 2.